Faithful out arguments

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49012 Making ops c10 full: out overloads with default arguments
* #49008 Making ops c10-full: Dimname arguments
* #49007 Making ops c10-full: ops blocked by manual registrations
* #49006 Making ops c10 full: ops returning multiple out arguments
* #48912 Make out ops c10-full (with hacky-wrapper)
* #48911 hacky_wrapper_for_legacy_signatures reorders out arguments
* #48910 Metaprogramming facilities to support hacky_wrapper to reorder function arguments
* #47793 Make aten::div.out c10-full
* #47792 VariableType calls faithful C++ API for c10-full out ops
* **#47712 Faithful out arguments**
* #47711 Enable faithful API for all ops

This adds a faithful API for ops with out arguments, as described in https://docs.google.com/document/d/1h7nBibRwkRLQ8rsPhfALlwWR0QbkdQm30u4ZBwmaps8/edit# .

After this, an op will generate the following overloads for the C++ API:

```cpp
// Generated from the aten::abs operator (NOT from aten::abs.out)
Tensor at::abs(Tensor& self)

// Generated from the aten::abs.out operator
Tensor& at::abs(Tensor& self, Tensor& out)
Tensor& at::abs_out(Tensor& out, Tensor& self)

```

This is an important step towards making those ops c10-full (it allows VariableType, XLA and other backends to ignore reordering and just call through with the same argument order), but this does not make any of those ops c10-full yet.
It enables the faithful API independent from c10-fullness. That means the API is more consistent with the same API for all ops and making an op c10-full in the future will not trigger future C++ API changes.

The old API is still around and valid, so backward compatibility is not broken there.
However, this does break bc in an a bit more unexpected way (in the same way as adding a new operator overload always breaks bc). Concretely, the `Tensor::m(at::my_op, args...)` only works if `at::my_op` doesn't have multiple overloads. Previously, out overloads were excluded from that rule and an operator that had one regular and one out overload still worked. This does now not work anymore because out overloads behave more like regular overloads. This also breaks `torch::nn::Functional(at::sigmoid)` for those ops. Furthermore, this will break any third party code that uses `at::my_op` as a symbol to get its function pointer without specifying the full argument list.

Those broken APIs, however, only worked for ops without overloads previously and adding any overload would have broken them. A workaround for users to still use those APIs is to explicitly specify the argument list like `torch::nn::Functional(static_cast<Tensor(*)(const Tensor&)>(at::sigmoid))`. But this is super ugly.
We probably should reconsider those APIs and reimplement them differently.

**edit**: To avoid breaking the APIs, we decided to introduce `at::myop_outf` overloads instead that represent the faithful version of out overloads. This means we don't do the API cleanup yet that this was planning, but there's plans to do that down the road the right way.

Differential Revision: [D24835252](https://our.internmc.facebook.com/intern/diff/D24835252/)

smessmer committed:
Faithful out arguments

This adds a faithful API for ops with out arguments, as described in https://docs.google.com/document/d/1h7nBibRwkRLQ8rsPhfALlwWR0QbkdQm30u4ZBwmaps8/edit# .

After this, an op will generate the following overloads for the C++ API:

```cpp
// Generated from the aten::abs operator (NOT from aten::abs.out)
Tensor at::abs(Tensor& self)

// Generated from the aten::abs.out operator
Tensor& at::abs(Tensor& self, Tensor& out)
Tensor& at::abs_out(Tensor& out, Tensor& self)

```

This is an important step towards making those ops c10-full (it allows VariableType, XLA and other backends to ignore reordering and just call through with the same argument order), but this does not make any of those ops c10-full yet.
It enables the faithful API independent from c10-fullness. That means the API is more consistent with the same API for all ops and making an op c10-full in the future will not trigger future C++ API changes.

Differential Revision: [D24835252](https://our.internmc.facebook.com/intern/diff/D24835252/)

[ghstack-poisoned]

bhosmer commented on tools/codegen/api/cpp.py:
```diff
@@ -252,16 +252,20 @@ def argument_not_this(
 
 def argument(
     a: Union[Argument, TensorOptionsArguments, ThisArgument],
+    is_out_argument: bool,
```
Should this not be a property (maybe synthetic) of `Argument`? It seems odd to have it in a sidecar like this - e.g. it's nonsense for it to be `True` if `a` is a `TensorOptionsArguments` or `ThisArgument`.

bhosmer commented on tools/codegen/api/cpp.py:
```diff
@@ -23,9 +23,9 @@
 # collisions, but functions are fair game to collide
 
-def name(func: FunctionSchema) -> str:
+def name(func: FunctionSchema, *, use_suffix_for_out_overloads: bool = True) -> str:
```
Seems like this should be a property (maybe synthetic) of `FunctionSchema`

ezyang replied:
It can't be part of FunctionSchema because we need to get both the faithful (`add`) and the unfaithful (`add_out`) name for a single FunctionSchema, there is no FunctionSchema variation. What this does suggest, though, was I was wrong to have individual functions for extracting name here, and we should go full on with CppSignature (which will avoid this very strange usage edge).

bhosmer commented on tools/codegen/api/cpp.py:
```diff
@@ -252,16 +252,20 @@ def argument_not_this(
+    is_out_argument: bool,
 ) -> Union[CppSingleArgumentPack, CppThisArgumentPack]:
     if isinstance(a, ThisArgument):
+        assert not is_out_argument
```
evidence :)

bhosmer commented on tools/codegen/api/cpp.py:
```diff
@@ -252,16 +252,20 @@ def argument_not_this(
+    is_out_argument: bool,
 ) -> CppArgumentPack:
     if isinstance(a, TensorOptionsArguments):
+        assert not is_out_argument
```
evidence :)

bhosmer commented on tools/codegen/api/cpp.py:
```diff
@@ -252,16 +252,20 @@ def argument_not_this(
 
 def argument_faithful(
     a: Union[Argument, TensorOptionsArguments, ThisArgument],
+    is_out_argument: bool,
```
see above

bhosmer commented on tools/codegen/api/cpp.py:
```diff
@@ -270,16 +274,15 @@ def argument_faithful(
 def group_arguments(
     func: FunctionSchema, *, method: bool
-) -> Sequence[Union[Argument, TensorOptionsArguments, ThisArgument]]:
+) -> Tuple[Sequence[Union[Argument, TensorOptionsArguments, ThisArgument]], Sequence[Argument]]:
```
At this point the returned structure should really become a class. The motivation for the imprecise list=of-unions representation was that it left the structure simple and flat, but with this tuple that's no longer the case, and we're going down a slippery slope here (tuple-stapling more and more stuff onto return values without defining a structure for them is like issue #1 when functional programs get too big IME). 

We should bite the bullet and make a `GroupedArguments` data class that models things properly. In the interests of expediency you could probably just do `GroupedArguments` with these two members, but see [this comment thread](https://github.com/pytorch/pytorch/pull/45890#discussion_r503646917) for more.

ezyang replied:
I wanted to show that the refactor in question wasn't so hard; it took me a few hours but now FunctionSchema has grouped arguments in https://github.com/pytorch/pytorch/pull/48195 . I'd like to see this PR reworked on top of it.

bhosmer commented on tools/codegen/api/dispatcher.py:
```diff
@@ -47,31 +47,34 @@ def returns_type(rs: Sequence[Return]) -> str:
     return cpp.returns_type(rs)
 
-def argument(a: Argument) -> DispatcherArgument:
+def argument(a: Argument, is_out_argument: bool) -> DispatcherArgument:
```
Same comment about `is_out_argument` here and below

bhosmer commented on tools/codegen/api/python.py:
```diff
@@ -439,7 +439,7 @@ def _cpp_signature(f: NativeFunction, *, method: bool = False) -> CppSignature:
 def has_tensor_options(f: NativeFunction) -> bool:
     return any(filter(lambda a: isinstance(a, TensorOptionsArguments),
-                      cpp.group_arguments(f.func, method=False)))
+                      cpp.group_arguments(f.func, method=False)[0]))
```
evidence of the need for `GroupedArguments` or similar

bhosmer commented on tools/codegen/gen.py:
```diff
@@ -334,10 +337,39 @@ def go(f: NativeFunction) -> Optional[str]:
         assert target is Target.DEFINITION
 
-        def generate_defn(sig: CppSignature) -> str:
+        def generate_defn(sig: CppSignature, faithful: bool) -> str:
```
`faithful` should be a property of `CppSignature`, no?

Edit: now that I've read the logic below, this seems even more straightforwardly true. `generate_defn` should be a simple consumer of a `CppSignature`. Any structuring of the signature itself should have already happened by the time we get here.

ezyang replied:
Yes, agreed here. @smessmer, I think we did talk about this in chat, but I think I did not completely understand the problem you were posing me. I think the relevant comment from you was:

> So if I go with your approach and group the arguments into regular and out arguments, then the CppSignature doesn't have that info anymore but now needs to store an additional flag of if the out arguments are supposed to come before or after the regular arguments

And so to solve this problem, I proposed "where you continue to maintain a list, but then you have fingers into particular pieces of the rep. So CppSignatures will be a list of CppArgument, but then it also has an out_argument list which redundantly specifies which of these arguments were the out arguments so you can conveniently reference them in translation".

So you did the marking. But you failed to maintained the correctly ordered list of arguments in CppSignature so that you don't need the flag here anymore! (And it seems to me like maintaining direct pointers to out arguments, rather than flagging them on the arguments themselves, is easier, because then you don't need to do the partitioning business here.)

bhosmer commented on tools/codegen/gen.py:
```diff
@@ -334,10 +337,39 @@ def go(f: NativeFunction) -> Optional[str]:
+            #    non-faithful signature has out arguments in the front, the dispatcher for c10-full ops takes them
+            #    in the back. We need to reorder them to the back.
+            # Cases 1, 2 don't need special handling and cases 3, 4 are detected and handled in the following code block.
+            # Note that cases 2, 3 will die as soon as all ops are c10-full but 1, 4 will remain.
```
This seems off - if we need to undo an upstream change, it means this logic should live upstream and produce different artifacts for this code to consume. Off the top of my head an obvious approach to do would be to produce a different `CppSignature` for the faithful and non-faithful cases. Is there an issue with doing that?

bhosmer reviewed (commented):
A few things I think we need to address here - interestingly they all have the same overall shape, moving new information into the data model and using it upstream, rather than adding free floating flag variables and using them downstream to tweak output :) happy to talk through any of these suggestions, not sure how much overlap there is between what I'm saying here and what you and @ezyang have talked about.

ezyang commented:
@smessmer Google docs are not visible to external people (because the FB gdocs org doesn't allow public sharing). You'll have to move it to your personal Google, or republish to https://github.com/pytorch/rfcs/ for accessibility

ezyang commented on tools/codegen/api/types.py:
```diff
@@ -211,6 +214,7 @@ def defn(self, name: Optional[str] = None, *, prefix: str = "") -> str:
+        out_arguments: Sequence[Argument],
         *,
         faithful: bool
     ) -> 'CppSignature':
```
Non-PR related philosophizing here: one of the consequences of immutable data types is that there is an equivalence between "data structure" and "memoized computation"; that is to say, if we have a constructor that does some nontrivial computation on data, we can choose to store the input data and then force clients to do the recompute every time, or we can upfront do the compute and then make it available. There isn't really any distinction and whether or not to slide from one end of the spectrum to the other is a mostly a matter of efficiency and clarity. (I guess this is why laziness is so good when you are purely functional).

bhosmer replied:
Another dimension that plays here is fanout - if there's a canonical representation and a bunch of useful projections, that often puts a natural boundary between what's stored and what's computed. 

Things get blurry when there's a good canonical representation, *but* a single useful projection kind of dominates. (I'm thinking of something like in the structured signature situation where most clients just want to iterate over some argument supertype, but the canonical representation would be much more precise, correct by construction etc.)

Argh SNIPED

ezyang commented:
Here's my new high level thinking about where we should be headed with the code generation here.

* The shared model has to be structured enough to cover the cross-product of all use-cases for each of the downstream APIs. Previously, this wasn't true, because we didn't do grouping in the model; but post https://github.com/pytorch/pytorch/pull/48195 it's now true. So now you can basically read out CppArguments straight from FunctionSchema, as long as you know if you are faithful and if you know you're a method/function.
* Given this, I suggest we STOP doing intermediate translations to `_argument_packs` and `_returns_type` in `CppSignature`; instead, the idea is that `CppSignature` *only* stores "is faithful? is method?" and we compute CppArguments on the fly when necessary. We can get rid of the CppArgumentPacks now since they're no longer necessary, just read out the structure from the model.
* And finally, quoting you from our personal chat, we can now do "schema -> exprs" directly. If you need to query some relevant information about how exactly cpp API was encoded, you on the fly compute the CppArgument, and then read out its name there. This solves my other constraint, "It should be possible to rename the name of a local variable (e.g., an argument binding) by modifying only one location in the code, and no others".

smessmer committed:
Update on "Faithful out arguments"


This adds a faithful API for ops with out arguments, as described in https://docs.google.com/document/d/1h7nBibRwkRLQ8rsPhfALlwWR0QbkdQm30u4ZBwmaps8/edit# .

After this, an op will generate the following overloads for the C++ API:

```cpp
// Generated from the aten::abs operator (NOT from aten::abs.out)
Tensor at::abs(Tensor& self)

// Generated from the aten::abs.out operator
Tensor& at::abs(Tensor& self, Tensor& out)
Tensor& at::abs_out(Tensor& out, Tensor& self)

```

This is an important step towards making those ops c10-full (it allows VariableType, XLA and other backends to ignore reordering and just call through with the same argument order), but this does not make any of those ops c10-full yet.
It enables the faithful API independent from c10-fullness. That means the API is more consistent with the same API for all ops and making an op c10-full in the future will not trigger future C++ API changes.

Differential Revision: [D24835252](https://our.internmc.facebook.com/intern/diff/D24835252/)

[ghstack-poisoned]

smessmer committed:
Update on "Faithful out arguments"


This adds a faithful API for ops with out arguments, as described in https://docs.google.com/document/d/1h7nBibRwkRLQ8rsPhfALlwWR0QbkdQm30u4ZBwmaps8/edit# .

After this, an op will generate the following overloads for the C++ API:

```cpp
// Generated from the aten::abs operator (NOT from aten::abs.out)
Tensor at::abs(Tensor& self)

// Generated from the aten::abs.out operator
Tensor& at::abs(Tensor& self, Tensor& out)
Tensor& at::abs_out(Tensor& out, Tensor& self)

```

This is an important step towards making those ops c10-full (it allows VariableType, XLA and other backends to ignore reordering and just call through with the same argument order), but this does not make any of those ops c10-full yet.
It enables the faithful API independent from c10-fullness. That means the API is more consistent with the same API for all ops and making an op c10-full in the future will not trigger future C++ API changes.

Differential Revision: [D24835252](https://our.internmc.facebook.com/intern/diff/D24835252/)

[ghstack-poisoned]

smessmer commented:
I reimplemented this PR on top of #48195. This makes it much better. PTAL

smessmer committed:
Update on "Faithful out arguments"


This adds a faithful API for ops with out arguments, as described in https://docs.google.com/document/d/1h7nBibRwkRLQ8rsPhfALlwWR0QbkdQm30u4ZBwmaps8/edit# .

After this, an op will generate the following overloads for the C++ API:

```cpp
// Generated from the aten::abs operator (NOT from aten::abs.out)
Tensor at::abs(Tensor& self)

// Generated from the aten::abs.out operator
Tensor& at::abs(Tensor& self, Tensor& out)
Tensor& at::abs_out(Tensor& out, Tensor& self)

```

This is an important step towards making those ops c10-full (it allows VariableType, XLA and other backends to ignore reordering and just call through with the same argument order), but this does not make any of those ops c10-full yet.
It enables the faithful API independent from c10-fullness. That means the API is more consistent with the same API for all ops and making an op c10-full in the future will not trigger future C++ API changes.

Differential Revision: [D24835252](https://our.internmc.facebook.com/intern/diff/D24835252/)

[ghstack-poisoned]

smessmer committed:
Update on "Faithful out arguments"


This adds a faithful API for ops with out arguments, as described in https://docs.google.com/document/d/1h7nBibRwkRLQ8rsPhfALlwWR0QbkdQm30u4ZBwmaps8/edit# .

After this, an op will generate the following overloads for the C++ API:

```cpp
// Generated from the aten::abs operator (NOT from aten::abs.out)
Tensor at::abs(Tensor& self)

// Generated from the aten::abs.out operator
Tensor& at::abs(Tensor& self, Tensor& out)
Tensor& at::abs_out(Tensor& out, Tensor& self)

```

This is an important step towards making those ops c10-full (it allows VariableType, XLA and other backends to ignore reordering and just call through with the same argument order), but this does not make any of those ops c10-full yet.
It enables the faithful API independent from c10-fullness. That means the API is more consistent with the same API for all ops and making an op c10-full in the future will not trigger future C++ API changes.

The old API is still around and valid, so backward compatibility is not broken there.
However, this does break bc in an a bit more unexpected way (in the same way as adding a new operator overload always breaks bc). Concretely, the `Tensor::m(at::my_op, args...)` only works if `at::my_op` doesn't have multiple overloads. Previously, out overloads were excluded from that rule and an operator that had one regular and one out overload still worked. This does now not work anymore because out overloads behave more like regular overloads.

Differential Revision: [D24835252](https://our.internmc.facebook.com/intern/diff/D24835252/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/test/basic.cpp:
```diff
@@ -288,7 +288,7 @@ void TestIndexingMixedDevice(DeprecatedTypeProperties& type) {
 void TestDispatch() {
   Tensor tensor = randn({20, 20});
   Tensor other = randn({20, 20});
-  auto result = tensor.m(relu).m(mse_loss, other, at::Reduction::Mean);
```
good bye m, it was nice knowing ya

ezyang commented on tools/codegen/api/cpp.py:
```diff
@@ -273,10 +273,11 @@ def argument_faithful(
         return argument(a)
 
 def group_arguments(
```
tbh this isn't really grouping arguments anymore, it's just flattening arguments into the appropriate order for cpp apis.

smessmer replied:
true, that could be renamed. It't independent from this PR though. I'll delay this as part of the follow-up clean-up you were planning to get rid of the grouping classes as well.

smessmer committed:
Update on "Faithful out arguments"


This adds a faithful API for ops with out arguments, as described in https://docs.google.com/document/d/1h7nBibRwkRLQ8rsPhfALlwWR0QbkdQm30u4ZBwmaps8/edit# .

After this, an op will generate the following overloads for the C++ API:

```cpp
// Generated from the aten::abs operator (NOT from aten::abs.out)
Tensor at::abs(Tensor& self)

// Generated from the aten::abs.out operator
Tensor& at::abs(Tensor& self, Tensor& out)
Tensor& at::abs_out(Tensor& out, Tensor& self)

```

This is an important step towards making those ops c10-full (it allows VariableType, XLA and other backends to ignore reordering and just call through with the same argument order), but this does not make any of those ops c10-full yet.
It enables the faithful API independent from c10-fullness. That means the API is more consistent with the same API for all ops and making an op c10-full in the future will not trigger future C++ API changes.

The old API is still around and valid, so backward compatibility is not broken there.
However, this does break bc in an a bit more unexpected way (in the same way as adding a new operator overload always breaks bc). Concretely, the `Tensor::m(at::my_op, args...)` only works if `at::my_op` doesn't have multiple overloads. Previously, out overloads were excluded from that rule and an operator that had one regular and one out overload still worked. This does now not work anymore because out overloads behave more like regular overloads. This also breaks `torch::nn::Functional(at::sigmoid)` for those ops. Furthermore, this will break any third party code that uses `at::my_op` as a symbol to get its function pointer without specifying the full argument list.

Those broken APIs, however, only worked for ops without overloads previously and adding any overload would have broken them. A workaround for users to still use those APIs is to explicitly specify the argument list like `torch::nn::Functional(static_cast<Tensor(*)(const Tensor&)>(at::sigmoid))`. But this is super ugly.
We probably should reconsider those APIs and reimplement them differently.

Differential Revision: [D24835252](https://our.internmc.facebook.com/intern/diff/D24835252/)

[ghstack-poisoned]

bhosmer commented on test/cpp/api/modules.cpp:
```diff
@@ -1451,7 +1453,7 @@ TEST_F(ModulesTest, FunctionalWithTorchFunction) {
 TEST_F(ModulesTest, FunctionalArgumentBinding) {
   auto functional =
-      Functional(torch::elu, /*alpha=*/1, /*scale=*/0, /*input_scale=*/1);
+      Functional(static_cast<Tensor(*)(const Tensor&, Scalar, Scalar, Scalar)>(torch::elu), /*alpha=*/1, /*scale=*/0, /*input_scale=*/1);
```
Per the PR description, this potentially kind of a rough change if people are using this idiom much. Did we consider using a different suffix for faithful-out? It's ugly but I'm not sure that matters.

bhosmer commented on tools/codegen/api/types.py:
```diff
@@ -150,49 +150,36 @@ class CppSignature:
-        faithful: bool
-    ) -> 'CppSignature':
-        if faithful:
+        positional_and_kwargs: List[Union[Argument, SelfArgument, TensorOptionsArguments]] = \
```
This next block looks like it was copy pasted from `cpp.group_arguments()`, that can't be the best way to do this can it?

bhosmer commented on tools/codegen/gen.py:
```diff
@@ -462,10 +462,15 @@ def __call__(self, f: NativeFunction) -> Optional[str]:
+            else:
+                sig = sig_group.signature
+
+            dispatcher_exprs = dispatcher.cpparguments_exprs(f.func, method=False, api_is_faithful=faithful)
```
Let's go thru this one together - it's worth working back from here to the things it's calling and figuring out how faithfulness can be better encapsulated.

smessmer replied:
as discussed offline, part of the complexity here comes from the fact that we need to match 2x2=4 scenarios. The API can be faithful/unfaithful and the dispatcher calling convention can also be faithful/unfaithful. This will get cleaner once the dispatcher calling convention is always faithful.

bhosmer reviewed (commented):
Some comments inline, I'll follow up 1-1 :)

ezyang commented on test/cpp/api/modules.cpp:
```diff
@@ -1451,7 +1453,7 @@ TEST_F(ModulesTest, FunctionalWithTorchFunction) {
 
 TEST_F(ModulesTest, FunctionalArgumentBinding) {
   auto functional =
-      Functional(torch::elu, /*alpha=*/1, /*scale=*/0, /*input_scale=*/1);
```
I guess this is why you work-chatted me about overloads being busted; not just `m()` stops working, `Functional()` *explicitly* stops working entirely (although, I probably would have suggested you rewrite these to pass in a lambda rather than casting to identify the overload instead). IDK man... I know you've invested a lot into this approach, but it really isn't a huge lift to just setup a completely new namespace, and the problems here fail the sniff test of being intrinsically related to the change you're trying to make (you're *just* trying to simplify codegen in autograd layer on redispatch and the sniff test says that unrelated UX shouldn't be a blocker here).

smessmer replied:
yeah that's why I reached out to you in workchat, having noticed that Functional breaks too, it's suddenly worse than I thought. I'll introduce `_outf` overloads for now and investigate the struct idea I had to fix this more long term.

ezyang commented on tools/codegen/api/dispatcher.py:
```diff
@@ -137,7 +137,22 @@ def cppargument_exprs(
+    else:
+        argument_packs = tuple(
+            cpp.argument(a) for a in arguments
+        )
```
You do grouping (with faithfulness) and then mapping between argument and `argument_faithful` here, but then you replicate the logic again at https://github.com/pytorch/pytorch/pull/47712/files#diff-2cccf17f15442a35105f349b3bebaf199294d8c3380767b73117031589832634R191 This seems a bit smelly to me.

smessmer replied:
the reason is that one thing branches based on if the dispatcher calling convention is faithful and the other thing branches based on if the API is faithful. Those are different things, we're actually handling 2x2=4 scenarios here.

ezyang commented on tools/codegen/api/types.py:
```diff
@@ -201,17 +188,31 @@ def _from_grouped_arguments(
+        return argument_packs
+
+    def name(self) -> str:
+        return cpp.name(self.func, use_suffix_for_out_overloads=not self.faithful)
```
Maybe it would be more direct and clear to just replace `use_suffix_for_out_overloads` with `faithful`.

ezyang commented on tools/codegen/api/types.py:
```diff
@@ -150,49 +150,36 @@ class CppSignature:
-        faithful: bool
-    ) -> 'CppSignature':
-        if faithful:
+        positional_and_kwargs: List[Union[Argument, SelfArgument, TensorOptionsArguments]] = \
```
This segment is something like a copy paste of `group_arguments` but with some tweaks, right? I do think copy pasting is the right call here, but because I think at some point we should delete `group_arguments` entirely. Doesn't have to be this PR though.

ezyang replied:
For this PR, it seems like a good idea to compare and contrast this to `group_arguments`

smessmer replied:
yeah this is actually doing exactly what `group_arguments` is doing. Seems calling into `group_arguments` is a trivial simplification.

ezyang commented on tools/codegen/api/dispatcher.py:
```diff
@@ -137,7 +137,22 @@ def cppargument_exprs(
+
+    return _cpparguments_exprs(argument_packs)
+
+def _cpparguments_exprs(args: Sequence[CppArgumentPack]) -> Sequence[DispatcherExpr]:
```
Not for this PR, but I hope you can see that CppArgumentPack is totally unnecessary now: the only difference is whether or not you construct a CppTensorOptionsArgumentPack or not, but we can just get rid of CppArgumentPack entirely and make sure we pass in faithful (which controls whether or not we constructed a CppTensorOptionsArgumentPack or not). This should make a number of things simpler.

ezyang reviewed (approved):
Still a bit nervous about m() and Functional(), but this seems good enough to go in.

bhosmer reviewed (changes requested):
Putting a changes requested to make sure we discuss the overloading issue (`Functional()` etc.), its possible downstream consequences and the cost of alternatives. Let's schedule a chat once you're settled.

The less consequential stuff (per earlier inline comments) we can potentially sort out post-PR, in conjunction with @ezyang's refactoring ideas.

smessmer committed:
Update on "Faithful out arguments"


This adds a faithful API for ops with out arguments, as described in https://docs.google.com/document/d/1h7nBibRwkRLQ8rsPhfALlwWR0QbkdQm30u4ZBwmaps8/edit# .

After this, an op will generate the following overloads for the C++ API:

```cpp
// Generated from the aten::abs operator (NOT from aten::abs.out)
Tensor at::abs(Tensor& self)

// Generated from the aten::abs.out operator
Tensor& at::abs(Tensor& self, Tensor& out)
Tensor& at::abs_out(Tensor& out, Tensor& self)

```

This is an important step towards making those ops c10-full (it allows VariableType, XLA and other backends to ignore reordering and just call through with the same argument order), but this does not make any of those ops c10-full yet.
It enables the faithful API independent from c10-fullness. That means the API is more consistent with the same API for all ops and making an op c10-full in the future will not trigger future C++ API changes.

The old API is still around and valid, so backward compatibility is not broken there.
However, this does break bc in an a bit more unexpected way (in the same way as adding a new operator overload always breaks bc). Concretely, the `Tensor::m(at::my_op, args...)` only works if `at::my_op` doesn't have multiple overloads. Previously, out overloads were excluded from that rule and an operator that had one regular and one out overload still worked. This does now not work anymore because out overloads behave more like regular overloads. This also breaks `torch::nn::Functional(at::sigmoid)` for those ops. Furthermore, this will break any third party code that uses `at::my_op` as a symbol to get its function pointer without specifying the full argument list.

Those broken APIs, however, only worked for ops without overloads previously and adding any overload would have broken them. A workaround for users to still use those APIs is to explicitly specify the argument list like `torch::nn::Functional(static_cast<Tensor(*)(const Tensor&)>(at::sigmoid))`. But this is super ugly.
We probably should reconsider those APIs and reimplement them differently.

Differential Revision: [D24835252](https://our.internmc.facebook.com/intern/diff/D24835252/)

[ghstack-poisoned]

bhosmer reviewed (approved):
Cool - very happy to dodge the newly-overloaded ops risk.
* couple nits inline
* after this stack lands we should do a refactor based on Ed's "high level thinking" comment, which will ripple into here in a good way
* I'll add an H1 item for the struct-overload-bundle idea

smessmer committed:
Update on "Faithful out arguments"


This adds a faithful API for ops with out arguments, as described in https://docs.google.com/document/d/1h7nBibRwkRLQ8rsPhfALlwWR0QbkdQm30u4ZBwmaps8/edit# .

After this, an op will generate the following overloads for the C++ API:

```cpp
// Generated from the aten::abs operator (NOT from aten::abs.out)
Tensor at::abs(Tensor& self)

// Generated from the aten::abs.out operator
Tensor& at::abs(Tensor& self, Tensor& out)
Tensor& at::abs_out(Tensor& out, Tensor& self)

```

This is an important step towards making those ops c10-full (it allows VariableType, XLA and other backends to ignore reordering and just call through with the same argument order), but this does not make any of those ops c10-full yet.
It enables the faithful API independent from c10-fullness. That means the API is more consistent with the same API for all ops and making an op c10-full in the future will not trigger future C++ API changes.

The old API is still around and valid, so backward compatibility is not broken there.
However, this does break bc in an a bit more unexpected way (in the same way as adding a new operator overload always breaks bc). Concretely, the `Tensor::m(at::my_op, args...)` only works if `at::my_op` doesn't have multiple overloads. Previously, out overloads were excluded from that rule and an operator that had one regular and one out overload still worked. This does now not work anymore because out overloads behave more like regular overloads. This also breaks `torch::nn::Functional(at::sigmoid)` for those ops. Furthermore, this will break any third party code that uses `at::my_op` as a symbol to get its function pointer without specifying the full argument list.

Those broken APIs, however, only worked for ops without overloads previously and adding any overload would have broken them. A workaround for users to still use those APIs is to explicitly specify the argument list like `torch::nn::Functional(static_cast<Tensor(*)(const Tensor&)>(at::sigmoid))`. But this is super ugly.
We probably should reconsider those APIs and reimplement them differently.

**edit**: To avoid breaking the APIs, we decided to introduce `at::myop_outf` overloads instead that represent the faithful version of out overloads. This means we don't do the API cleanup yet that this was planning, but there's plans to do that down the road the right way.

Differential Revision: [D24835252](https://our.internmc.facebook.com/intern/diff/D24835252/)

[ghstack-poisoned]

smessmer committed:
Update on "Faithful out arguments"


This adds a faithful API for ops with out arguments, as described in https://docs.google.com/document/d/1h7nBibRwkRLQ8rsPhfALlwWR0QbkdQm30u4ZBwmaps8/edit# .

After this, an op will generate the following overloads for the C++ API:

```cpp
// Generated from the aten::abs operator (NOT from aten::abs.out)
Tensor at::abs(Tensor& self)

// Generated from the aten::abs.out operator
Tensor& at::abs(Tensor& self, Tensor& out)
Tensor& at::abs_out(Tensor& out, Tensor& self)

```

This is an important step towards making those ops c10-full (it allows VariableType, XLA and other backends to ignore reordering and just call through with the same argument order), but this does not make any of those ops c10-full yet.
It enables the faithful API independent from c10-fullness. That means the API is more consistent with the same API for all ops and making an op c10-full in the future will not trigger future C++ API changes.

The old API is still around and valid, so backward compatibility is not broken there.
However, this does break bc in an a bit more unexpected way (in the same way as adding a new operator overload always breaks bc). Concretely, the `Tensor::m(at::my_op, args...)` only works if `at::my_op` doesn't have multiple overloads. Previously, out overloads were excluded from that rule and an operator that had one regular and one out overload still worked. This does now not work anymore because out overloads behave more like regular overloads. This also breaks `torch::nn::Functional(at::sigmoid)` for those ops. Furthermore, this will break any third party code that uses `at::my_op` as a symbol to get its function pointer without specifying the full argument list.

Those broken APIs, however, only worked for ops without overloads previously and adding any overload would have broken them. A workaround for users to still use those APIs is to explicitly specify the argument list like `torch::nn::Functional(static_cast<Tensor(*)(const Tensor&)>(at::sigmoid))`. But this is super ugly.
We probably should reconsider those APIs and reimplement them differently.

**edit**: To avoid breaking the APIs, we decided to introduce `at::myop_outf` overloads instead that represent the faithful version of out overloads. This means we don't do the API cleanup yet that this was planning, but there's plans to do that down the road the right way.

Differential Revision: [D24835252](https://our.internmc.facebook.com/intern/diff/D24835252/)

[ghstack-poisoned]
